### PR TITLE
[WIP] Add dependency injection functionality, Closes #85

### LINKED ISSUE #85: Dependency Injection Annotation
Provide a way for the command manager to receive dependencies and provide standard dependency injection abilities

Such as

```java
manager.addDependency(Plugin.class, plugin);
manager.addDependency(MyPlugin.class, plugin);
manager.addDependency(MyObjectManager.class, objManager);

public MyCommandClass extends BaseCommand {
    @Dependency MyPlugin plugin;
    @Dependency MyObjectManager manager;
    // ... commands
}
```

On registration, ACF can fulfill the dependencies as provided, as commonly seen in more featured dependency injection frameworks.

### PR DESCRIPTION
**Pull request by MiniDigger**
also ran some intellij inspections.

### CODE REVIEW

**Comment by MiniDigger** [comment]
possible improvements:
[ ] more default dependencies
[ ] injecting of subtypes, currently only exactly matching types are injected
[ ] allow injection of classes that were not registered via a @Singleton like functionality

File: bukkit/src/main/java/co/aikar/commands/BukkitCommandContexts.java [hunk 1]
```diff
@@ -74,7 +74,7 @@ public BukkitCommandContexts(BukkitCommandManager manager) {
 
                 throw new InvalidCommandArgument(false);
             }
-            return players.toArray(new OnlinePlayer[players.size()]);
```
Comment (aikar) [inline bukkit/src/main/java/co/aikar/commands/BukkitCommandContexts.java]: Undo this change

File: core/src/main/java/co/aikar/commands/CommandContexts.java [hunk 2]
```diff
@@ -174,7 +174,7 @@
                 ACFUtil.sneaky(new IllegalStateException("Weird Command signature... String[] should be last or @Split"));
             }
 
-            String[] result = args.toArray(new String[args.size()]);
+            String[] result = args.toArray(new String[0]);
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandContexts.java]: Undo this change

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 3]
```diff
@@ -377,6 +380,71 @@ public void addSupportedLanguage(Locale locale) {
...
+     * Registers an instance of a class to be registered as an injectable dependency.<br>
+     * The command manager will attempt to inject all fields in a command class that are annotated with
+     * {@link co.aikar.commands.annotation.Dependency} with the provided instance.
+     *
+     * @param clazz the class the injector should look for when injecting
+     * @param instance the instance of the class that should be injected
+     * @throws IllegalArgumentException when instance isn't an instance of clazz
+     * @throws IllegalStateException when there is already an instance for the provided class registered
+     */
+    public void registerDependency(Class<?> clazz, Object instance){
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: Use <T> with Class <T> clazz, T instance, to require it actually satisfies the class its providing for.

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 4]
```diff
@@ -377,6 +380,71 @@ public void addSupportedLanguage(Locale locale) {
...
+     * The command manager will attempt to inject all fields in a command class that are annotated with
+     * {@link co.aikar.commands.annotation.Dependency} with the provided instance.
+     *
+     * @param clazz the class the injector should look for when injecting
+     * @param key the key which needs to be present if that
+     * @param instance the instance of the class that should be injected
+     * @throws IllegalArgumentException when instance isn't an instance of clazz
+     * @throws IllegalStateException when there is already an instance for the provided class registered
+     */
+    public void registerNamedDependency(Class<?> clazz, String key, Object instance){
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: Same as above for <T>, but just make this registerDependency, since its a different signature

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 5]
```diff
@@ -377,6 +380,71 @@ public void addSupportedLanguage(Locale locale) {
...
+        if(!clazz.isInstance(instance)){
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: Not necessary with <T> restrictions

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 6]
```diff
@@ -377,6 +380,71 @@ public void addSupportedLanguage(Locale locale) {
...
+            throw new IllegalArgumentException(instance.getClass().getName() + " isn't a subclass of " + clazz.getName() + "!");
+        }
+        if(dependencies.containsRow(clazz) && dependencies.containsColumn(key)){
+            throw new IllegalStateException("There is already an instance of " + clazz.getName() + " with the key " + key + " registered!");
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: Not sure we should throw, but just let it overwrite, maybe return old, like a typical put operation.

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 7]
```diff
@@ -377,6 +380,71 @@ public void addSupportedLanguage(Locale locale) {
...
+        dependencies.put(clazz, key, instance);
+    }
+
+    /**
+     * Attempts to inject instances of classes registered with {@link CommandManager#registerDependency(Class, Object)}
+     * into all fields that are marked with {@link Dependency}.
+     *
+     * @param baseCommand the instance which fields should be filled
+     */
+    public void inject(BaseCommand baseCommand){
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: make package private

File: core/src/main/java/co/aikar/commands/RegisteredCommand.java [hunk 8]
```diff
@@ -201,7 +201,7 @@ void handleException(CommandIssuer sender, List<String> args, Exception e) {
     @Nullable
     Map<String, Object> resolveContexts(CommandIssuer sender, List<String> args, int argLimit) throws InvalidCommandArgument {
         args = Lists.newArrayList(args);
-        String[] origArgs = args.toArray(new String[args.size()]);
+        String[] origArgs = args.toArray(new String[0]);
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/RegisteredCommand.java]: Undo change

File: sponge/src/main/java/co/aikar/commands/SpongeCommandContexts.java [hunk 9]
```diff
@@ -119,7 +117,7 @@ public SpongeCommandContexts(final SpongeCommandManager manager) {
 
                 throw new InvalidCommandArgument(false);
             }
-            return players.toArray(new OnlinePlayer[players.size()]);
+            return players.toArray(new OnlinePlayer[0]);
```
Comment (aikar) [inline sponge/src/main/java/co/aikar/commands/SpongeCommandContexts.java]: Undo change

File: core/src/main/java/co/aikar/commands/CommandManager.java (hunk 3 above)
Comment (aikar) [reply core/src/main/java/co/aikar/commands/CommandManager.java]: generic T, github stripped it...

Comment (aikar) [reply core/src/main/java/co/aikar/commands/CommandManager.java]:
```java
public <T> void registerDependency(Class<T> clazz, T instance) {
```

File: core/src/main/java/co/aikar/commands/CommandManager.java (hunk 7 above)
Comment (aikar) [reply core/src/main/java/co/aikar/commands/CommandManager.java]: I think also more descriptive "injectDependencies"

File: core/src/main/java/co/aikar/commands/CommandManager.java (hunk 3 above)
Comment (MiniDigger) [reply core/src/main/java/co/aikar/commands/CommandManager.java]: I had this and it didn't work. when I want to pass the plugins class, I can't cast to it, thus throwing a compiler error. https://i.imgur.com/ynHskze.png

**Comment by MiniDigger** [comment]
if some bungee or sponge ppl could suggest default dependencies or the platform that would be great.
suggestions for bukkit are welcome too.

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 10]
```diff
@@ -377,6 +380,66 @@ public void addSupportedLanguage(Locale locale) {
...
+                if(object == null){
+                    throw new UnresolvedDependencyException("Could not find a registered instance of " +
+                            field.getType().getName() + " with key " + key + " for field " + field.getName() +
+                            " in class " + baseCommand.getClass().getName());
+                }
+
+                try {
+                    field.setAccessible(true);
+                    field.set(baseCommand, object);
+                    field.setAccessible(false);
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: Don't know if theres a reason to set this to false? what if it was already accessible before.

File: core/src/main/java/co/aikar/commands/CommandManager.java [hunk 11]
```diff
@@ -377,6 +380,66 @@ public void addSupportedLanguage(Locale locale) {
...
+    }
+
+    /**
+     * Attempts to inject instances of classes registered with {@link CommandManager#registerDependency(Class, Object)}
+     * into all fields that are marked with {@link Dependency}.
+     *
+     * @param baseCommand the instance which fields should be filled
+     */
+    void injectDependencies(BaseCommand baseCommand){
+        for (Field field : baseCommand.getClass().getDeclaredFields()) {
```
Comment (aikar) [inline core/src/main/java/co/aikar/commands/CommandManager.java]: If getDeclaredFields is the one that only includes its own fields, then switch to the other. we want to include parent class fields too, but I'm not at a java doc atm.

File: sponge/src/main/java/co/aikar/commands/SpongeCommandManager.java [hunk 12]
```diff
@@ -67,6 +67,10 @@ public SpongeCommandManager(PluginContainer plugin) {
         this.formatters.put(MessageType.INFO, new SpongeMessageFormatter(TextColors.BLUE, TextColors.DARK_GREEN, TextColors.GREEN));
         this.formatters.put(MessageType.HELP, new SpongeMessageFormatter(TextColors.AQUA, TextColors.GREEN, TextColors.YELLOW));
         getLocales(); // auto load locales
+
+        //TODO more default dependencies for sponge
+        registerDependency(plugin.getClass(), plugin);
+        registerDependency(PluginContainer.class, plugin);
```
Comment (aikar) [inline sponge/src/main/java/co/aikar/commands/SpongeCommandManager.java]: Remove the PluginContainer one since per @kashike its not the same thing.

File: core/src/main/java/co/aikar/commands/CommandManager.java (hunk 11 above)
Comment (MiniDigger) [reply core/src/main/java/co/aikar/commands/CommandManager.java]: when we don't use getDeclaredFields, we can't inject private fields. do we want to just iterate over both?
